istioctl/cmd/validation.go

```diff
@@ -0,0 +1,414 @@
+			namespaceForTheTempPod, maxPodWaitDuration)
+		if err != nil {
+			return nil, fmt.Errorf("error when waiting for the pod: %v", err)
+		}
+		execCmd := "cat " + certPath
+		execOutput, err := podExec(namespaceForTheTempPod, nameForTheTempPod, containerNameForTheTempPod,
+			execCmd, true, kubeconfig)
+		if err != nil {
+			return nil, fmt.Errorf("error when exec on the pod: %v", err)
+		}
+		if execOutput == "" {
+			return nil, fmt.Errorf("empty pod exec output")
+		}
+		caCert = []byte(execOutput)
+	}
+	b, _ := pem.Decode(caCert)
+	if b == nil {
+		return nil, fmt.Errorf("could not decode pem")
+	}
+	if b.Type != "CERTIFICATE" {
+		return nil, fmt.Errorf("ca certificate contains wrong type: %v", b.Type)
+	}
+	if _, err := x509.ParseCertificate(b.Bytes); err != nil {
+		return nil, fmt.Errorf("ca certificate parsing returns an error: %v", err)
+	}
+
+	return caCert, nil
+}
+
+func createPod(k8sInterface kubernetes.Interface, podName, namespace, podLabelKey,
+	podLabelVal, containerName, containerImage string, containerCmd []string) error {
+	req := &corev1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   podName,
+			Labels: map[string]string{podLabelKey: podLabelVal},
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				{
+					Name:    containerName,
+					Image:   containerImage,
+					Command: containerCmd,
+				},
+			},
+		},
+	}
+	resp, err := k8sInterface.CoreV1().Pods(namespace).Create(req)
+	if err != nil {
+		return err
+	}
+	if resp == nil {
+		return fmt.Errorf("pod creation response is nil")
+	}
+	return nil
+}
+
+func waitForServerRunning(host string, port int, maxWaitTime time.Duration) error {
+	startTime := time.Now()
+	timerCh := time.After(maxWaitTime - time.Since(startTime))
+	for {
+		err := isTLSServerRunning(host, port)
+		if err == nil {
+			fmt.Printf("the server at %v:%v is running\n", host, port)
+			break
+		}
+		fmt.Printf("the server at %v:%v is not running: %v\n", host, port, err)
+		select {
+		case <-timerCh:
+			return fmt.Errorf("server is not running within %v", maxWaitTime)
+		default:
+			time.Sleep(2 * time.Second)
+		}
+	}
+	return nil
+}
+
+func waitForPodRunning(client kubernetes.Interface, podName, podLabelKey, podLabelVal, namespace string,
+	maxWaitTime time.Duration) error {
+	fieldSelector := fields.SelectorFromSet(map[string]string{
+		"metadata.name":      podName,
+		"metadata.namespace": namespace}).String()
+	labelSelector := k8slabels.SelectorFromSet(map[string]string{podLabelKey: podLabelVal}).String()
+	listOptions := metav1.ListOptions{
+		FieldSelector: fieldSelector,
+		LabelSelector: labelSelector,
+	}
+	watch, err := client.CoreV1().Pods(namespace).Watch(listOptions)
+	if err != nil {
+		return fmt.Errorf("failed to set up a watch for pod (error: %v)", err)
+	}
+	events := watch.ResultChan()
```

Comment:
Seems racy, what happens if the pod is already created at this point?

Reply:
Done: this function has been removed.